### ISSUE TITLE
Support for easy migration between self-compiled and official releases

### ISSUE THREAD

**Issue by cpfeiffer** [Freeyourgadget/Gadgetbridge#945]
When you always use the fdroid version of Gadgetbridge, everything is fine. As soon as you self-compile Gadgetbridge, it will be signed with debug keys, and you will need to uninstall the previous version before you can install the self-compiled version.

Uninstalling will delete the database, the preferences and everything else, including the directory where the db exports live. So even if you performed an export, all your settings and data will be gone after uninstallation.

Long story short, we need a mechanism to switch between differently signed versions. If we export the data to a different place, where other apps (including the differently signed Gadgetbridge) can read it, we may want to encrypt the whole export with a passphrase, so that another rogue app could not decipher it.

Other apps solve this with storing everything externally "in the cloud", so the nextcloud integration might also offer a way to achieve this.

**Comment by TaaviE** [comment]
I proposed a change https://github.com/Freeyourgadget/Gadgetbridge/pull/949 to modify the package name a bit for debug builds. I think it's a good way to have way less hassle if one wants to test debug alongside release. Sharing configurations should be arranged somehow too between debug and release, but I haven't put much thought into that.

**Comment by IzzySoft** [comment]
Not being a dev, I can only report from "hear-say" and reading: "shared preferences" are probably out as different signatures are used. But if, as suggested by Taavi, packages also have different names, they could offer a method to have their data accessed protected by a specific (self-declared) permission (where the name must be adjusted and should be bound to the package name, as the same permission cannot be declared twice).

As an additional protection, a call to such "access method" should trigger a "warning" by the app providing the data, to have the user confirm it – to prevent "malicious authors" of sneaking in that permission into their apps and stealthily steal our data: Export would only happen on explicit approval, with rejecting access on timeout (such a transfer would usually be triggered manually anyway if I understand correctly.

**Comment by TaaviE** [comment]
@IzzySoft This could work. I can imagine the permission management system to be reused for BG JS and network-related stuff. Hmm, what do the primary developers think though?

**Comment by IzzySoft** [comment]
@TaaviE also an idea, yes: storing "network requests" in a kind of "backlog table" and broadcasting some "GB_NETWORK_JOB_RECEIVED" event other apps could respond to, polling stuff "tagged" for them and coming back with result code or content. But for that there shouldn't be approval needed each time, as that should work without user interaction – so the user should just have to approve the app once (but then only for this kind of activity, i.e. polling the "task table" and responding to those requests, not for "general access").

All together this would center around an API provided by GB. And what we're talking about here now goes beyond what this issue was created for :wink:
